### DISCUSSION TITLE
Casting multi-dimensional arrays to one-dimensional arrays

### DISCUSSION THREAD

**Discussion by runny-yolk** [dotnet/csharplang#6403] [Language Ideas] [answered]
Hi,

I feel like it should be possible to cast a multi-dimensional array (eg. `int[,]`) to a one dimensional array (eg. `int[]`).

My understanding is that, in memory, a multidimensional array is basically a single dimensional array, all that happens is when you initialise a multidimensional array `var a = new int[width, height]`, the length of the backing one dimensional array is `width * height`, and when you use the indexer method `a[x, y]`, a calculation is done to get the relevant one dimensional index that looks something like `x * height + y`.

In short, multi/one dimensional arrays are the same thing under the hood (as I understand it), just multidimensional arrays provide a wrapper to read/write the data in a more convenient way for some use cases, so theoretically, you should be able to cast from one to the other.

One of these use cases where multidimensional arrays are convenient is doing dynamic mesh generation in Unity. Triangles are groups of 3 indices, so for example, if you were building a mesh with 5 triangles, you'd make an array like `var tris = new int[5, 3]`.

However, you run into a problem, which is that the `Mesh` class requires a one dimensional array for its `triangles` field, but because you can't cast from multi to one dimensional arrays, you have to do a copy with `Array` or `Buffer` `.Copy`, and for my use case (and in game development generally) you want to save on performance as much as possible, so doing a copy isn't ideal, and casting is just easier and cleaner.

It's niche, but it'd be nice to have, my current solution is a custom multidimensional wrapper which feels weird given C# does have multidimensional arrays built in.

Thanks for reading!

**Comment by ufcpp** [comment]
https://stackoverflow.com/a/73373888/16828735

**Comment by huoyaoyuan** [comment, marked as answer]
In current CLR implementation, SZ (Single dimension, zero-based) array and MD array are handled quite differently. Their header layouts are different. SZ array implements the collection interfaces while MD array doesn't.
It may not be possible to implement all the features on MD array, but using them as spans would be straight achievable. See also https://github.com/dotnet/runtime/issues/49269 https://github.com/dotnet/runtime/issues/27690 .

**Reply by runny-yolk** [reply]
Ah, thanks for the info! Sounds like this is the closest things could get to what I want with the current CLR implementation.
